pkg/backends/calico/client.go

```diff
@@ -809,30 +817,35 @@ func (c *client) OnUpdates(updates []api.Update) {
 			log.Info("Starting route generator due to service advertisement update")
 			var err error
 			if c.rg, err = NewRouteGenerator(c); err != nil {
-				log.WithError(err).Error("Failed to start route generator, unable to advertise services")
+				log.WithError(err).Error("Failed to start route generator, unable to advertise node-specific service routes")
 				c.rg = nil
 			} else {
 				c.rg.Start()
 			}
 		}
 
+		// Update external IP CIDRs. In v1 format, they are a single comma-separated
+		// string. If the string isn't empty, split on the comma and pass a list of strings
+		// to the route generator.  An empty string indicates a withdrawal of that set of
+		// service IPs.
+		if len(c.cache["/calico/bgp/v1/global/svc_external_ips"]) > 0 {
+			c.onExternalIPsUpdate(strings.Split(c.cache["/calico/bgp/v1/global/svc_external_ips"], ","))
+		}
+
+		// Same for cluster CIDRs.
+		if len(c.cache["/calico/bgp/v1/global/svc_cluster_ips"]) > 0 {
+			c.onClusterIPsUpdate(strings.Split(c.cache["/calico/bgp/v1/global/svc_cluster_ips"], ","))
+		}
+
 		if c.rg != nil {
-			// Update CIDRs. In v1 format, they are a single comma-separate string. If the string isn't empty,
-			// split on the comma and pass a list of strings to the route generator.
-			// An empty string indicates a withdrawal of that set of service IPs.
-			var externalIPs []string
-			if len(c.cache["/calico/bgp/v1/global/svc_external_ips"]) > 0 {
-				externalIPs = strings.Split(c.cache["/calico/bgp/v1/global/svc_external_ips"], ",")
-			}
-			var clusterCIDRs []string
-			if len(c.cache["/calico/bgp/v1/global/svc_cluster_ips"]) > 0 {
-				clusterCIDRs = strings.Split(c.cache["/calico/bgp/v1/global/svc_cluster_ips"], ",")
+			// Trigger the route generator to recheck and advertise or withdraw
+			// node-specific routes.
+			select {
```

Comment:
Worth moving the select into a method of the route cache.  `c.rg.TriggerResync()` is a nice clean API where the channel machinery becomes an implementation detail.